### DISCUSSION TITLE
I need to output the position and label of the detected object

### DISCUSSION THREAD

**Discussion by yidaoyu** [ultralytics/ultralytics#1179] [Q&A] [answered]
yolo8 is a very good image recognition model, but now I need to output the position and label of the detected object. How can I do this?
I found this in yolo8's documentation, I try to use these to help me get more information about the target, but probs, masks return none

model = YOLO("mybest.pt") # load a pretrained model (recommended for training)
results = model("mypath") # predict on an image

for result in results:
boxes = result.boxes.xyxy # Boxes object for bbox outputs
print(boxes)
masks = result.masks # Masks object for segmenation masks outputs
print(masks) # none
probs = result.probs # Class probabilities for classification outputs
print(probs) # none

**Comment by Laughing-q** [comment]
@yidaoyu hi, if your model is a detection model then it's normal that `masks` and `probs` is None.

**Reply by yidaoyu** [reply]
Thanks. You know that when outputting the result, there is not only a prediction box, but also a label on the box to indicate what object it is. Now I need this label and the position of the box, so what should I do?

**Reply by glenn-jocher** [reply]
Hello @yidaoyu! To obtain the labels along with the positions of the bounding boxes, you can access the `pred` attribute from the results. This attribute contains the predicted bounding boxes along with the class IDs and confidence scores. Here's how you can use it:

```python
results = model("mypath")  # predict on an image
preds = results.pred  # Get predictions

for pred in preds:
    for *box, conf, cls in pred:
        label = results.names[int(cls)]  # Get the label of the class
        print(f"Label: {label}, Box: {box}, Confidence: {conf}")
```

This will print out the label, the position of the bounding box (as x1, y1, x2, y2), and the confidence score for each detected object. The `results.names` attribute maps class IDs to their respective labels. Keep in mind that the bounding box coordinates are in the format (x1, y1, x2, y2), where (x1, y1) is the top-left corner and (x2, y2) is the bottom-right corner of the bounding box.

**Reply by mohnish-bhadra** [reply]
Hi, @glenn-jocher  when I am using the code you have provided, this is the output I am receiving:

 preds = results.pred  # Get predictions
AttributeError: 'list' object has no attribute 'pred'

What do I do to fix this

**Reply by glenn-jocher** [reply]
Hello @mohnish-bhadra,

Thank you for reaching out! It looks like there was a misunderstanding in the code example. The `results` object is indeed a list of `Results` objects, and you should access the attributes directly from each `Results` object in the list. Here's the corrected approach:

```python
results = model("mypath")  # predict on an image

for result in results:
    boxes = result.boxes.xyxy  # Bounding box coordinates
    confs = result.boxes.conf  # Confidence scores
    classes = result.boxes.cls  # Class IDs

    for box, conf, cls in zip(boxes, confs, classes):
        label = result.names[int(cls)]  # Get the label of the class
        print(f"Label: {label}, Box: {box.tolist()}, Confidence: {conf.item()}")
```

This will give you the label, position of the bounding box, and the confidence score for each detected object.

If you continue to experience issues, please ensure you are using the latest version of the Ultralytics YOLO package. If the problem persists, providing a minimum reproducible example would be very helpful for further diagnosis. You can find guidance on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions! 😊

**Comment by Laughing-q** [comment, marked as answer]
@yidaoyu `result.boxes.cls` is the predicted label corresponding to the boxes.

**Reply by yidaoyu** [reply]
I tried using result.boxes.cls, but he returns
cls: tensor([0.], device='cuda:0')

**Reply by Laughing-q** [reply]
that's the cls index which means the category of this object is the first class you defined.

**Reply by yidaoyu** [reply]
Thank you very much, your answer reminded me, I returned to my data.yaml file, found the class name I wrote in the name line, and the object label matched the returned result.boxes.cls, can yolo8 Show my class name (corresponding to the name of data.yaml), of course, it doesn’t matter if it’s not there, I can write this class name in my python code, it would be better if yolo8 can show it

**Reply by Laughing-q** [reply]
you can easily get the plotted image which shows the class name by:
```python
for result in results:
    plotted_img = result.plot()
```

**Reply by yidaoyu** [reply]
wow，I can't wait to try this code: plotted_img = result.plot(), but his return is plotted: [[[37 35 24]
   [17 16 6]
   [33 30 25]
   ...
   [19 19 19]
   [38 38 38]
   [19 19 19]]
i didn't understand this part

**Reply by Laughing-q** [reply]
ah that's the plotted image, so you only want to access the predicted class name corresponding to the boxes?
```python
for result in results:
    cls = result.boxes.cls
    for c in cls:
        print(result.names[int(c)])
```

**Reply by yidaoyu** [reply]
awesome, that's it, thank you very much (although said many times), but all my problems were solved, this helped me a lot

**Reply by glenn-jocher** [reply]
@yidaoyu you're very welcome! I'm glad to hear that your issues have been resolved. If you have any more questions or need further assistance in the future, don't hesitate to reach out. Happy coding with YOLOv8! 😊🚀
